v2/pkg/founat/fou.go

```diff
@@ -228,22 +236,138 @@ func (t *fouTunnel) addPeer6(addr net.IP) (netlink.Link, error) {
 		LinkAttrs:  attrs,
 		Ttl:        225,
 		EncapType:  netlink.FOU_ENCAP_DIRECT,
 		EncapDport: uint16(t.port),
-		EncapSport: uint16(t.port),
+		EncapSport: encapSport,
 		Remote:     addr,
 		Local:      t.local6,
 	}
 	if err := netlink.LinkAdd(link); err != nil {
 		return nil, fmt.Errorf("netlink: failed to add fou6 link: %w", err)
 	}
 
+	if err := setupFlowBasedIP6TunDevice(); err != nil {
+		return nil, fmt.Errorf("netlink: failed to setup ipip device: %w", err)
+	}
+
 	return netlink.LinkByName(linkName)
 }
 
+// setupFlowBasedIP[4,6]TunDevice creates an IPv4 or IPv6 tunnel device
+//
+// This flow based IPIP tunnel device is used to decapsulate packets from
+// the router Pods.
+//
+// Calling this function may result in tunl0 (v4) or ip6tnl0 (v6)
+// fallback interface being renamed to coil_tunl or coil_ip6tnl.
+// This is to communicate to the user that this plugin has taken
+// control of the encapsulation stack on the netns, as it currently
+// doesn't explicitly support sharing it with other tools/CNIs.
+// Fallback devices are left unused for production traffic.
+// Only devices that were explicitly created are used.
+//
+// This fallback interface is present as a result of loading the
+// ipip and ip6_tunnel kernel modules by fou tunnel interfaces.
+// These are catch-all interfaces for the ipip decapsulation stack.
+// By default, these interfaces will be created in new network namespaces,
+// but this behavior can be disabled by setting net.core.fb_tunnels_only_for_init_net = 2.
+func setupFlowBasedIP4TunDevice() error {
+	ipip4Device := "coil_ipip4"
+	// Set up IPv4 tunnel device if requested.
+	if err := setupDevice(&netlink.Iptun{
+		LinkAttrs: netlink.LinkAttrs{Name: ipip4Device},
+		FlowBased: true,
+	}); err != nil {
+		return fmt.Errorf("creating %s: %w", ipip4Device, err)
+	}
+
+	// Rename fallback device created by potential kernel module load after
+	// creating tunnel interface.
+	if err := renameDevice("tunl0", "coil_tunl"); err != nil {
+		return fmt.Errorf("renaming fallback device %s: %w", "tunl0", err)
+	}
+
+	return nil
+}
+
+// See setupFlowBasedIP4TunDevice
+func setupFlowBasedIP6TunDevice() error {
+	ipip6Device := "coil_ipip6"
+
+	// Set up IPv6 tunnel device if requested.
+	if err := setupDevice(&netlink.Ip6tnl{
+		LinkAttrs: netlink.LinkAttrs{Name: ipip6Device},
+		FlowBased: true,
+	}); err != nil {
+		return fmt.Errorf("creating %s: %w", ipip6Device, err)
+	}
+
+	// Rename fallback device created by potential kernel module load after
+	// creating tunnel interface.
+	if err := renameDevice("ip6tnl0", "coil_ip6tnl"); err != nil {
+		return fmt.Errorf("renaming fallback device %s: %w", "tunl0", err)
+	}
+
+	return nil
+}
+
+// setupDevice creates and configures a device based on the given netlink attrs.
+func setupDevice(attrs netlink.Link) error {
+	name := attrs.Attrs().Name
+
+	// Reuse existing tunnel interface created by previous runs.
+	l, err := netlink.LinkByName(name)
+	if err != nil {
```

Comment:
I think it should be restricted to `netlink.LinkNotFoundError` and other errors should be returned.

Reply:
Done, thanks for your comment